webapp/src/ts/modules/contacts/contacts-edit.component.ts

```diff
@@ -221,6 +228,21 @@ export class ContactsEditComponent implements OnInit, OnDestroy, AfterViewInit {
     return formId;
   }
 
+  private async validateParent(parentId, childTypeId) {
```

Comment:
These are just suggestions, so take what makes sense to you and leave the rest.

- If a function is called 'validate...' I assume it is going to _fail_ for invalid data (instead of just return `false`).
- IMHO the `LineageModelGeneratorService` is overkill here since we don't care about hydrating the lineage of the parent.  Can just pull in the `ChtDatasourceService` instead.

```ts
  private async validateParentForCreateForm() {
    const getPlace = await this.chtDatasourceService.bind(Place.v1.get);
    const parent = await getPlace({ uuid: this.contact.parent });
    const parentType = this.contactTypesService.getTypeId(parent);
    const validParentType = !this.contact.parent || parentType;
    const isValidChildType = () => this.contactTypesService
      .getChildren(parentType)
      .then((validChildTypes) => validChildTypes.some(({ id }) => id === this.contact.contact_type));
    if (!validParentType || !await isValidChildType()) {
      throw new Error(`Cannot create a ${this.contact.contact_type} as a child of a ${parentType}.`);
    }
  }
```

Reply:
I've gone with `ChtDatasourceService`, thanks for the suggestion.  But I went with the imperative mode, which I find easier to read and write.